### ISSUE TITLE
Model Conversion and Optimization

### ISSUE THREAD

**Issue by WhiteMasky** [Darcyddx/Active-Vision#39]
- [ ]  Convert the trained TensorFlow models (e.g., MobileNetV2, RNN models) to TensorFlow Lite format to ensure they can run on mobile devices.
- [ ]  Optimize the TensorFlow Lite models by using built-in ops (BUILTINS) to improve inference speed on Android and other devices.
- [ ]  Perform FP16 quantization to reduce the model size and improve inference speed.

**Comment by WhiteMasky** [comment]
model = tf.keras.models.load_model('tennis_rnn.h5')
```                                                                                                                                               | Reconstructs the network architecture in code, then loads weights into it:  
```python
model = Model(..., name="tennis_rnn_unrolled")
model.load_weights("tennis_rnn.h5")
```                                                                                          |
| **RNN Implementation**  | Uses the default GRU (unroll=False), which under the hood creates TensorArray ops (TensorList) for the recurrence.                                                     | Sets `unroll=True` on the GRU layer, expanding the recurrence into a fixed computation graph and eliminating any TensorArray use.                              |
| **Converter Settings**  | - Supports both TFLITE_BUILTINS and TensorFlow native ops (SELECT_TF_OPS)  
- Disables the experimental TensorList downgrade  
```python
converter.target_spec.supported_ops = [
  tf.lite.OpsSet.TFLITE_BUILTINS,
  tf.lite.OpsSet.SELECT_TF_OPS
]
converter._experimental_lower_tensor_list_ops = False
``` | Uses the default converter settings (only built-in TFLite ops), with no special flags or delegates required:  
```python
converter = tf.lite.TFLiteConverter.from_keras_model(model)
```                                                         |
| **Quantization**        | Enables dynamic-range quantization (optional):  
```python
converter.optimizations = [tf.lite.Optimize.DEFAULT]
```                                                                                                        | Leaves quantization commented out (produces a full-precision model):  
```python
# converter.optimizations = [tf.lite.Optimize.DEFAULT]
```                                                           |
| **Output File**         | `model_with_select_tf_ops.tflite` — this file contains Select TF Ops and therefore requires the Select TF Ops delegate at runtime.                                      | `tennis_rnn_unrolled.tflite` — a pure BUILTINS-only model that runs on standard TFLite without extra delegates.                                              |
| **Pros & Cons**         | **Pros:**  
- No model code rewrite needed  
- Keeps original graph structure  
**Cons:**  
- Requires TF Select Ops delegate (larger binary, extra dependency)                                                                             | **Pros:**  
- Produces a pure TFLite-only model deployable on any standard TFLite runtime  
**Cons:**  
- Must manually maintain a separate “unrolled” model definition in code                                                                      |

**Summary**:  
- Use **tfops.py** if you want to convert your existing `.h5` model with minimal code changes, by enabling Select TF Ops and bypassing the TensorArray downgrade.  
- Use **static.py** if you need a standalone, built-in-ops-only TFLite model (no extra delegates) by unrolling your RNN layers at graph-build time.

**Comment by WhiteMasky** [comment]
| Aspect             | Details |
|--------------------|---------|
| **Error Cause**    | TensorFlow Lite conversion failed because `tf.TensorListReserve` operation requires a static `element_shape`, but a dynamic shape was detected. TensorList ops are complex for TFLite to lower without specific settings. |
| **Solution 1**     | Use **Select TF ops** by setting:<br>`converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS, tf.lite.OpsSet.SELECT_TF_OPS]`<br>and<br>`converter._experimental_lower_tensor_list_ops = False` |
| **Pros (Solution 1)** | - Easy to apply.<br>- Retains model correctness.<br>- No need to modify the model. |
| **Cons (Solution 1)** | - Larger model size.<br>- Dependency on TensorFlow runtime at inference. |
| **Solution 2**     | Modify model to avoid TensorList operations (e.g., replace dynamic list ops with static tensor allocations). |
| **Pros (Solution 2)** | - Fully TFLite-compatible model.<br>- Smaller model size.<br>- No extra TensorFlow runtime needed at inference. |
| **Cons (Solution 2)** | - Requires manual model redesign.<br>- Can be time-consuming.<br>- May affect model behavior if not careful. |
| **Solution 3**     | Try static shape enforcement during model export/training if possible (e.g., ensure tensor list shapes are constant). |
| **Pros (Solution 3)** | - Cleaner model design.<br>- Smooth TFLite conversion without extra ops. |
| **Cons (Solution 3)** | - May not be feasible depending on application requirements.<br>- Might involve training code changes. |

**Comment by WhiteMasky** [comment]
| Aspect             | Details |
|--------------------|---------|
| **Error Cause**    | TensorFlow Lite conversion failed because `tf.TensorListReserve` operation requires a static `element_shape`, but a dynamic shape was detected. TensorList ops are complex for TFLite to lower without specific settings. |
| **Solution 1**     | Use **Select TF ops** by setting:<br>`converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS, tf.lite.OpsSet.SELECT_TF_OPS]`<br>and<br>`converter._experimental_lower_tensor_list_ops = False` |
| **Pros (Solution 1)** | - Easy to apply.<br>- Retains model correctness.<br>- No need to modify the model. |
| **Cons (Solution 1)** | - Larger model size.<br>- Dependency on TensorFlow runtime at inference. |
| **Solution 2**     | Modify model to avoid TensorList operations (e.g., replace dynamic list ops with static tensor allocations). |
| **Pros (Solution 2)** | - Fully TFLite-compatible model.<br>- Smaller model size.<br>- No extra TensorFlow runtime needed at inference. |
| **Cons (Solution 2)** | - Requires manual model redesign.<br>- Can be time-consuming.<br>- May affect model behavior if not careful. |
| **Solution 3**     | Try static shape enforcement during model export/training if possible (e.g., ensure tensor list shapes are constant). |
| **Pros (Solution 3)** | - Cleaner model design.<br>- Smooth TFLite conversion without extra ops. |
| **Cons (Solution 3)** | - May not be feasible depending on application requirements.<br>- Might involve training code changes. |

**Comment by WhiteMasky** [comment]
Final decision: Use Select TF ops to convert into Float16 precision model. 

**Precision Type**: The input and output remain in Float32, but the model weights are compressed to Float16.

**Advantages**: Reduces model size by approximately 50% and enables faster inference, especially when using GPU/NPU acceleration. && Do not requires representative data for calibration

**Disadvantages**: Slight loss in accuracy.

**Use Case**: GPU inference on mobile devices, such as Android NNAPI or iOS Metal.
